protocol/src/credentials.rs

```diff
@@ -0,0 +1,84 @@
+use std::{borrow::Cow, fmt::Display, str::FromStr};
+
+use aes_siv::{Aes256SivAead, KeyInit};
+use anyhow::{ensure, format_err, Error};
+use base64::display::Base64Display;
+use base64::{prelude::BASE64_URL_SAFE_NO_PAD, Engine};
+use derivative::Derivative;
+use derive_more::AsRef;
+use generic_array::{typenum::U64, GenericArray};
+use rand::{distributions::Alphanumeric, distributions::DistString, rngs::OsRng};
+use serde::{de, Deserialize, Deserializer, Serialize, Serializer};
+
+#[derive(Clone, AsRef, Derivative, Deserialize, Serialize)]
+#[derivative(Debug)]
+pub struct AccessToken(#[derivative(Debug = "ignore")] String);
+
+const ACCESS_TOKEN_LENGTH: usize = 64;
+
+impl AccessToken {
+    pub fn generate() -> Self {
+        Self(Alphanumeric.sample_string(&mut OsRng, ACCESS_TOKEN_LENGTH))
+    }
+}
+
+impl FromStr for AccessToken {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        ensure!(
+            s.len() == ACCESS_TOKEN_LENGTH,
+            "invalid length; got {}, expected {ACCESS_TOKEN_LENGTH}",
+            s.len(),
+        );
+        Ok(Self(s.to_owned()))
+    }
+}
+
+#[derive(Clone, Derivative)]
+#[derivative(Debug)]
+pub struct EncryptionKey(#[derivative(Debug = "ignore")] GenericArray<u8, U64>);
+
+impl EncryptionKey {
+    pub fn generate() -> Self {
+        Self(Aes256SivAead::generate_key(&mut OsRng))
+    }
+
+    pub fn get(&self) -> &GenericArray<u8, U64> {
+        &self.0
+    }
+
+    /// We intentionally don't implement `Display` on `Self` to prevent accidental
+    /// encryption key disclosure (in logs, etc).
+    pub fn fmt(&self) -> impl Display + '_ {
```

Comment:
For `AccessToken`, instead of making a symmetric `fn display_unmasked(&self) → impl Display + '_`, I made `fn as_unmasked_str(&self) → &str`. `display_unmasked` would be ideal for specifying the token in HTTP headers (it accepts `impl Display`), but `impl sqlx::Encode` is needed to specify the token in SQL queries — it could be implemented, but it would force us to bring even more `sqlx` imports into the protocol.